### DISCUSSION TITLE
Hide a collection in the nav for a given role

### DISCUSSION THREAD

**Discussion by sarthakgaur** [directus/directus#11233] [Feature Requests]
I want a role to have read access to a collection, but hide it from their list of collections they can see on the sidebar.

Example:

Let's say there are two tables, `scores` and `exams`. The `exams` table contains all the details of exams, and `scores` table contains the marks scored by a student for a particular exam. Students need read access to the `exams` table because they need to see how much marks they scored for a particular exam.

I want students to see their scores, but I want to hide the `exams` table for them. Right now, if I hide a table it will hide it for all the roles, including the admin.

**Comment by rijkvanzanten** [comment]
Removing the `read` permission for that role also hides the collection in the admin app 👍🏻

**Reply by sarthakgaur** [reply]
I want to give a role read access to a table, but also hide the table for them.

Example:

`exams` table schema:
1. `title`

`scores` table schema:

1. `student`
2. `marks`
3. `exam`

A student needs read access to the `exams` table, otherwise they will not be able to see the `exam` field in the `scores` table. I am fine with giving them read access to the `exams` table. I just want to hide the `exams` table in the list of collections they can see on their sidebar. How can I do this?

**Reply by azrikahar** [reply]
Currently there isn't really a toggle for collections in navigation menu, but we can take advantage of the current setup where we can nest collections & folders in Data Model! The solution you can use would be:

1. Create a Folder in Data Model page
2. Nest `exams` table inside the newly created folder
3. Click on the right side menu of that newly created folder (three dots button) and click `Make Collection Hidden` to make that folder hidden

Result of above steps from Admin view:

![image](https://user-images.githubusercontent.com/42867097/150907387-f878ba9e-7f8a-4d98-9592-ba5a4c6565a3.png)

Now any collections under `Folder 1` will not be visible to any roles even if they have read permission to all collections/folders in that folder.

**Reply by sarthakgaur** [reply]
@azrikahar This does hide the `exams` table for students, and they are able to see the `exam` field in the `scores` table. But, this also hides all the collections inside the folder for the admin as well. Now, if I want to see the `exams` collection, I have to manually type `/admin/content/exams` in the URL.

**Reply by azrikahar** [reply]
It's true that it'll hide for admin, but you should be able to unhide it by right clicking the navigation sidebar and a context menu with `Show Hidden Collections` should appear. That said, seems like this is also not restricted to admin only 🤔 @rijkvanzanten is/should this feature intended for admin only?

**Reply by sarthakgaur** [reply]
@azrikahar Yes, a student can also right click the navigation sidebar and see all the hidden collections.

If admins can have ability to hide collections per role, that would be great. I know you guys get a lot of feature requests, but I think many people will find this feature useful, and it doesn't look hard to implement.

**Reply by benhaynes** [reply]
Just adding this to the conversation: I am not against a dedicated "LIST" permission for Collections in the Nav. This could be nested under VIEW, or standalone. In other words, the data (items) could be seen as different from the listing (Collections). 🙂

**Reply by abdonrd** [reply]
@benhaynes that sounds good!

**Reply by piotr-cz** [reply]
@benhaynes 
At first the *List* permission came to my mind, but I'm not sure it would solve this issue.

For example I'm creating collections that are primarily used in M2O fields as a configurable drop-downs. I'd like to hide them in Content area however these should still be available as lists to choose items from.

The Hidden collection feature is good as long as users don't get an idea to right-click on the sidebar.
In that case restricting functionality to *Admin Access* would do the job, but it's sort of breaking change.

**Comment by abdonrd** [comment]
I agree that it is an interesting function!

I have another use case:
- Profiles
- Countries

I want only certain roles to be able to see and create the Countries from the menu on the left.

But that certain roles that can view and create Profiles can assign a birth country (relationship) from the list of Countries.

**Reply by umityilmazz** [reply]
did you solve this problem? If you could solve it, how did you go about it?

**Reply by abdonrd** [reply]
No. 😅

**Reply by 1Luc1** [reply]
I don't know if it solves this special case, but i solved the initial question from this thread with a custom hook. Got the idea from here - [Martijn Boland Blog](https://blogs.taiga.nl/martijn/posts/use-a-hook-to-hide-modules-for-non-admin-roles-in-directus/).
Here is my adapted solution: https://github.com/1Luc1/directus-extension-hook-hide-module/blob/main/src/index.js

**Comment by claytongulick** [comment]
We have the same problem, we have some roles that need to see certain collections in the content app, and other users that shouldn't. 

This is orthogonal to permissions though - we may want the users to see the items in a list on a patient's record (episodes) but not see the "episodes" collection in the main nav.

Especially with the case of extensions, where roles need view access to a collection, but we don't want them to see it in the main nav.

One solution to the problem would be to add conditions to the collection level the same way they are on the field level.

We'd be interested in making this feature our next sponsored item @rijkvanzanten .

Thanks!!

**Reply by rijkvanzanten** [reply]
@claytongulick Would you want to be able to override the _structure_ as well, or only override the hidden status for collections? 🤔

**Reply by claytongulick** [reply]
I'm glad you mentioned that! Just ran into the issue yesterday where I needed to hide fields based on roles.

I thought I'd be able to do this with conditions, but $CURRENT_USER doesn't work in a key value.

Is that what you're thinking?

**Reply by w0kyj** [reply]
@claytongulick 
> I'm glad you mentioned that! Just ran into the issue yesterday where I needed to hide fields based on roles.
> 
> I thought I'd be able to do this with conditions, but $CURRENT_USER doesn't work in a key value.
> 
If you just want to hide a field for a role, you can just disable the fields for the collection permissions: Use Custom Permissions and deselect the Field Permissions --> Fields

> Is that what you're thinking?
No, we just want to confirm that you don't need to restructure the Collection Navigation tree order on a per role basis. Currently the Collection order is set in the Data Model. For this feature, want to ensure it is still okay to keep the global data model order and just add ability to hide a globally visible collection for specific roles.

Data Model:
table1 - visible
table2 - hidden
table3 - visible
table4 - visible

Feature as described:
Role:
Restricted Role1
table1 - visible
table2 - hidden
**table3 - hidden -- overrides global data model setting**
table4 - visible

Question asked is does this need to support overriding the Global Data Model tree order:
table1 - visible
**table4 - visible -- overrides global data model order**
table2 - hidden
**table3 - hidden -- overrides global data model setting**

Let us know your thoughts. 

NOTE: The reorder is a much larger feature

**Reply by claytongulick** [reply]
Thanks!

> If you just want to hide a field for a role, you can just disable the fields for the collection permissions: Use Custom Permissions and deselect the Field Permissions --> Fields

Unfortunately this doesn't work for me use case because I need the field to be accessible via the api and extensions, I just want to hide them in the UI for the user. Things like IDs for M2O fields that the user doesn't care about.

> Let us know your thoughts.

Thanks for the clarification!

For now, I think just having the ability to hide a collection based on a condition/role is all we need.

If condition keys supported $CURRENT_USER it could potentially work for both the collection hiding based on condition, as well as individual field hiding based on condition.

I was browsing through the source for the condition parsing and execution, it looks fairly straightforward to add $current_user to the key parser. Would it be helpful if I worked on a PR for that?

**Reply by rijkvanzanten** [reply]
> I was browsing through the source for the condition parsing and execution, it looks fairly straightforward to add $current_user to the key parser. Would it be helpful if I worked on a PR for that?

That'd be a great improvement to add 🙂

**Reply by Pachat** [reply]
Is there a way to deactivate "Show Hidden Collections"?

I tried to deactivate "Show Hidden Collections" for everyone by adding
```
div[data-v-26b985dc] {
    display: none;
}
```
to Custom CSS. But that deactivates all drop down menus in Roles. Have to find a better solution. Even hard coded, but did not find where to.

> > If you just want to hide a field for a role, you can just disable the fields for the collection permissions: Use Custom Permissions and deselect the Field Permissions --> Fields
> 
> Unfortunately this doesn't work for me use case because I need the field to be accessible via the api and extensions, I just want to hide them in the UI for the user. Things like IDs for M2O fields that the user doesn't care about.
> 
> > Let us know your thoughts.

At present, as I am in the same situation (M2O accessible but should be hidden from the UI), I'd just like the "Show Hidden Collections" to work only for admins.

**Reply by preetamslot** [reply]
For now, we can add a bit of CSS to hide the links in the navbar:
```
.v-list-item.link.hidden {
    display: none;
}
```

And if you want to hide the arrows:
```
.content-navigation .v-list-item-icon.activator-icon {
     display: none;
}

```

**Reply by piotr-cz** [reply]
Probably this trick could be used together with custom embeds: https://github.com/directus/directus/pull/16650

**Reply by preetamslot** [reply]
I did not know custom embeds were a thing, looks promising!

**Reply by preetamslot** [reply]
@piotr-cz  I could not see how to get the role info in the embed.  The only thing I can control at the moment is the domain, So I added an extra subdomain domain for admins. The editors use the original subdomain (cms.mydomain.com)

```
 embed(
    "body",
    () => `
   
    <script>
    document.addEventListener("DOMContentLoaded", function(event) {
          const element = document.getElementById("app");
          const subdomain = window.location.host.split('.').shift()
          if(subdomain == "cms") {
             element.classList.add("role-editor"); 
            }
    });
   
    </script>
    
    `
  );
```

The downside is that the admin domain cant use SSO and system emails will always point to the PUBLIC_URL.  it is far from perfect, but for my project it does what I need.

**Comment by Dominic-Marcelino** [comment]
From skimming the comments it looks like most users don't need any data-restrictions but only want to clean things up.

How about adding the user-role to a html root element and then adding the collection-name as class to the collection list? This would allow an admin to hide them simply via CSS.

**Reply by paescuj** [reply]
There's an existing discussion on this: https://github.com/directus/directus/discussions/10942

**Comment by rijkvanzanten** [comment]
Heya! Thanks for opening this feature request!

This feature request has received over **15 votes** from the community. This means we'll move this feature request to the Under Review state! 

The Core team will schedule a meeting to review this request as soon as possible. The discussion will then be approved or denied. You may or may not be invited to join this meeting with the core team.

_For more information, see our [Feature Request Process](https://docs.directus.io/contributing/feature-request-process.html)._

**Comment by ChristianDC13** [comment]
Hello, this feature has already received a significant number of votes, and it's clear that the community is eager to see something like this integrated. If it's not too much trouble, I would like to volunteer to work on the implementation. It would be a pleasure for me to contribute.

**Reply by rijkvanzanten** [reply]
There isn't enough information present in this thread for me to know how this is supposed to work.. The first step would be to define the RFC for this particular feature, so we can align on the list of requirements to call this a success 🙂

**Reply by Pachat** [reply]
Thank you for reviewing this highly requested feature.
The current issue stems from the UI layer being too tightly coupled with the data layer. This limitation reduces Directus's interface to primarily a development and testing tool, leading developers to use it mainly as an API backend in conjunction with other libraries like Nuxt.
To make the Directus UI more practical for direct use, we need features like this. 
The examples provided in this thread clearly demonstrate this necessity.
- Regarding implementation, the logical place would be within the Policies section, adding a Collection "**List**" column that would be enabled by default. When unchecked, the collection would be hidden for that specific policy.
- Additionally, we should add an "**Viewing hidden collections**" permission, which would be disabled by default.

**Reply by Ayo91** [reply]
I have created a feature request for the second point. Im aligned that should be a setting only for admin users: https://github.com/directus/directus/discussions/24559

The implementation is already done

**Comment by falmanna** [comment]
Are there are hacks, custom extensions, or suggestions on how to achieve this without changing permissions? 

We simply want to hide some collections from certain roles.

**Reply by 1Luc1** [reply]
I solved the initial question from this thread with a custom hook.
Got the idea from here - [Martijn Boland Blog](https://blogs.taiga.nl/martijn/posts/use-a-hook-to-hide-modules-for-non-admin-roles-in-directus/).
Here is my adapted solution: https://github.com/1Luc1/directus-extension-hook-hide-module/

**Reply by falmanna** [reply]
@1Luc1 Thanks, but seems like this is for modules. I want to apply the same logic to collections and collection groups. Any idea?

**Comment by cmann90** [comment]
We created a feature request and provided code for a solution where only admins can see "Show Hidden Collections".
Maybe that is a good intermediate solution for this more advanced feature request?

Please have a look and upvote it in case of interest:
#24559

**Comment by domdus** [comment]
Hi, I released this extension to cover a bunch requests for UI customization - including hiding of collections per role:
https://github.com/domdus/directus-extension-module-permissions

"Module Permissions" in the marketplace.
